### PR TITLE
Revert index scan

### LINKED ISSUE #183: Revert index scan functionality
@Tishj mentioned in #97 that we don't have tests for index scans. So I tried adding some basic tests for this. And as it turns out index scans are very much (and possibly completely) broken. This trivial example crashes the server for me:

```sql
CREATE TABLE test_table (id int primary key, name text);
SELECT * FROM test_table where id = 1;
```

### PR DESCRIPTION
**Pull request by mkaruza**
Reverting index scan feature as explained in #183. Removed query planning from execution. 

Three commits that should not be squashed when merging.

Fixes #183

### CODE REVIEW

File: src/catalog/pgduckdb_table.cpp [hunk 1]
```diff
@@ -36,7 +35,7 @@ PostgresTable::PopulateColumns(CreateTableInfo &info, Oid relid, Snapshot snapsh
 	auto tupleDesc = RelationGetDescr(rel);
 
 	if (!tupleDesc) {
-		elog(ERROR, "Failed to get tuple descriptor for relation with OID %u", relid);
```
Comment (JelteF) [inline src/catalog/pgduckdb_table.cpp:39]: I think this is probably a good change, but let's move it to a separate PR. It's nice to keep behavioral and refactoring changes separate, both for review and possible future `git revert`.

File: include/pgduckdb/catalog/pgduckdb_table.hpp [hunk 2]
```diff
@@ -52,20 +52,4 @@ class PostgresHeapTable : public PostgresTable {
...
-
-public:
-	// -- Table API --
-	unique_ptr<BaseStatistics> GetStatistics(ClientContext &context, column_t column_id) override;
-	TableFunction GetScanFunction(ClientContext &context, unique_ptr<FunctionData> &bind_data) override;
-	TableStorageInfo GetStorageInfo(ClientContext &context) override;
-
-private:
-	Path *path;
-	PlannerInfo *planner_info;
```
Comment (JelteF) [inline include/pgduckdb/catalog/pgduckdb_table.hpp:68]: Do we still need the PlannerInfo to be passed into the sequence scan now? i.e. I think we were only using it to choose between sequence scan and index scan, so I think we can get rid of storing it in the `"postgres_state"` key of `context->registered_state`.

Comment (Tishj) [reply include/pgduckdb/catalog/pgduckdb_table.hpp:68]: I think you're right, I think that might also remove the entire need for `PostgresContextState`, I believe `m_rtables` and `m_needed_columns` as well as `m_query_string` are not used ?

File: src/catalog/pgduckdb_table.cpp [hunk 3]
```diff
@@ -47,14 +47,25 @@ PostgresTable::PopulateColumns(CreateTableInfo &info, Oid relid, Snapshot snapsh
...
 		     duck_type.ToString().c_str());
 	}
 
 	RelationClose(rel);
 	return true;
 }
 
+Cardinality
+PostgresTable::GetTableCardinality(Oid relid) {
+	auto rel = RelationIdGetRelation(relid);
```
Comment (JelteF) [inline src/catalog/pgduckdb_table.cpp:60]: I think this probably needs the `DuckdbProcessLock`.

File: include/pgduckdb/catalog/pgduckdb_table.hpp (hunk 2 above)
Comment (JelteF) [reply include/pgduckdb/catalog/pgduckdb_table.hpp:68]: I'm also fine with doing that in a follow up PR btw.

**Review by JelteF — APPROVED** [review]
Apart from the missing process lock I think this can be merged. 

Like @Tishj said I think we can remove more code. But I'm also fine with that happening in a follow up PR.

File: src/catalog/pgduckdb_table.cpp (hunk 3 above)
Comment (mkaruza) [reply src/catalog/pgduckdb_table.cpp:60]: Let me give another pass through code; I'm seeing this `RelationIdGetRelation` used on few other places (unprotected by lock) maybe it is worth checking if this function can be used in only one function and than we would pass pointer to its data.

Comment (mkaruza) [reply src/catalog/pgduckdb_table.cpp]: @Tishj can you look into latest commit it should centralize opening/closing relation for execution

File: src/catalog/pgduckdb_table.cpp [hunk 4]
```diff
@@ -25,18 +27,29 @@ extern "C" {
...
 bool
-PostgresTable::SetTableInfo(CreateTableInfo &info, Oid relid, Snapshot snapshot) {
-	auto rel = RelationIdGetRelation(relid);
+PostgresTable::SetTableInfo(CreateTableInfo &info, ::Relation rel) {
 	auto tupleDesc = RelationGetDescr(rel);
 
 	if (!tupleDesc) {
-		elog(WARNING, "Failed to get tuple descriptor for relation with OID %u", relid);
+		elog(WARNING, "Failed to get tuple descriptor for relation with OID %u", rel->rd_id);
 		RelationClose(rel);
```
Comment (JelteF) [inline src/catalog/pgduckdb_table.cpp:53]: This one should now be removed I think.

**Review by JelteF — APPROVED** [review]
One small comment, but other than that this looks good now.

File: src/catalog/pgduckdb_table.cpp (hunk 4 above)
Comment (mkaruza) [reply src/catalog/pgduckdb_table.cpp:53]: Yes, thanks. I'm thinking if this check is even needed? If we have relation object opened before should we always have tuple desc information?

**Comment by JelteF** [comment]
Oh one more small note (I forgot it last time too when merging a PR without squashing). Can you update the commit messages to include the PR number at the end of the title in parenthesis, ie. ` (#247)`

File: src/catalog/pgduckdb_table.cpp (hunk 4 above)
Comment (JelteF) [reply src/catalog/pgduckdb_table.cpp:53]: Yeah, that sounds reasonable. There's no way for this to fail. It's a simple macro that reads a field:
```
#define RelationGetDescr(relation) ((relation)->rd_att)
```

**Review by JelteF — APPROVED** [review]

**Pull request merged**